services/src/main/java/io/grpc/protobuf/services/ProtoReflectionServiceV1.java

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright 2016 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.protobuf.services;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+import io.grpc.BindableService;
+import io.grpc.ExperimentalApi;
+import io.grpc.InternalServer;
+import io.grpc.Server;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1.ErrorResponse;
+import io.grpc.reflection.v1.ExtensionNumberResponse;
+import io.grpc.reflection.v1.ExtensionRequest;
+import io.grpc.reflection.v1.FileDescriptorResponse;
+import io.grpc.reflection.v1.ListServiceResponse;
+import io.grpc.reflection.v1.ServerReflectionGrpc;
+import io.grpc.reflection.v1.ServerReflectionRequest;
+import io.grpc.reflection.v1.ServerReflectionResponse;
+import io.grpc.reflection.v1.ServiceResponse;
+import io.grpc.stub.ServerCallStreamObserver;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.WeakHashMap;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ *
+ * <p>Separately tracks mutable and immutable services. Throws an exception if either group of
+ * services contains multiple Protobuf files with declarations of the same service, method, type, or
+ * extension.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+public final class ProtoReflectionServiceV1 extends ServerReflectionGrpc.ServerReflectionImplBase {
```

Comment:
As a follow-up PR, we'll want to remove the "extends ServerReflectionImplBase" from the public API surface. Yes, nobody can use it, but let's still not expose it. That is better as a follow-up, so that git will see the move/copy and then separately the changes made to the service.